datafusion/physical-expr/src/scalar_function.rs

```diff
@@ -243,6 +292,7 @@ pub fn create_physical_expr(
             Arc::new(fun.clone()),
             input_phy_exprs.to_vec(),
             return_type,
+            Arc::new(config_options.clone()),
```

Comment:
This line worries me -- it means each distinct scalar function in the plan will get an entirely new copy of the ConfigOptions.

I think the `Arc` should be passed in as the argument like this (I realize this will be a significant code change) so that the config options are copied at most once per plan

```rust
/// Create a physical expression for the UDF.
pub fn create_physical_expr(
    fun: &ScalarUDF,
    input_phy_exprs: &[Arc<dyn PhysicalExpr>],
    input_schema: &Schema,
    args: &[Expr],
    input_dfschema: &DFSchema,
    config_options: &Arc<ConfigOptions>, // <--- I think this should be an `Arc`
) -> Result<Arc<dyn PhysicalExpr>> {
```

I think that would also make it clearer that physical planning makes a read only copy of the configuration (`Arc<ConfigOptions>`) that is then unchanged during execution

Reply:
That was my main concern as well. I am unsure that this change should merged in as is to be honest as any fix will be just as disruptive as this PR is api wise.

Pushing the clone higher up the stack is possible but I did run into major issues trying to push it all the way up including all kinds of disruptive changes like changing OptimizerConfig from &ConfigOptions to the Arc version. Doing the opposite - pushing &ConfigOptions all the way down ran into issues with ScalarFunctionExpr and DynEq/DynHash.

Speaking for myself I prefer the latter as I feel it's less disruptive overall ... I just couldn't get it to work before. Changing the signature of DynHash and DynEq traits may allow it to work but I haven't tried that yet. Maybe with some more thought in January I can get it to work but I expect my January time to be pretty limited.

Reply:
> Pushing the clone higher up the stack is possible but I did run into major issues trying to push it all the way up including all kinds of disruptive changes like changing OptimizerConfig from &ConfigOptions to the Arc version. Doing the opposite - pushing &ConfigOptions all the way down ran into issues with ScalarFunctionExpr and DynEq/DynHash.

If you have an `Arc<ConfigOptions>` you can always get a `&ConfigOptions` by calling `options.as_ref()`

So in other words, I don't know if you need to push the Arc through everywhere (at least at first) -- we could just change the create physical expr. I would love to help with this -- perhaps I can find time later in the week (though I have a few other things going on too)

Another option to avoid cloning `ConfigOptions` for each instance of `ScalarFunction` might be  to add some sort of API / way for the function to communicate "I need the config options"